### ISSUE TITLE
[💡SUG] Improved MacridVAE

### ISSUE THREAD

**Issue by Nemexur** [RUCAIBox/RecBole#859]
**Is your feature request related to a problem? Please describe.**
There is an issue with MacridVAE implementation in prototype clustering. The training procedure considers sampling a probability simplex from Gumbel-Softmax, however, MacridVAE in RecBole samples a one-hot vector from Categorical distribution. It slightly differs from the authors' implementation. In addition, `D.OneHotCategorical` doesn't support backpropagation, whereas `tfp.RelaxedOneHotCategorical` in original implementation does.

**Describe the solution you'd like**
There is a simple solution to implement the required behavior in PyTorch.
```python
if self.nogb:
    cates = torch.softmax(cates_logits, dim=-1)
else:
    cates_sample = F.gumbel_softmax(cates_logits, tau=1, hard=False, dim=-1)
    cates_mode = torch.softmax(cates_logits, dim=-1)
    cates = (self.training * cates_sample + (1 - self.training) * cates_mode)
```

**Comment by 2017pxy** [comment]
@Nemexur Hi, thanks for your advice and we have updated the code in #861 .
